feat(apigateway): access logging (#6559)
aws/aws-cdk: packages/@aws-cdk/aws-apigateway/README.md

  }
})
```
### Access Logging

The following example will configure API Gateway to enable custom access logging

```ts
const api = new apigateway.RestApi(this, 'books', {
  deployOptions: {
    accessLogSetting: {

>         destinationArn: 'arn:aws:logs:us-east-1:123456789012:log-group:sample_group_name',

Reviewer: I think this should be:

 `destinationArn: new apigateway.CloudWatchLogsDestination('arn:aws:logs:us-east-1:123456789012:log-group:sample_group_name');`

Reviewer: Actually, on fiddling in my IDE, due to the interface input into `apigateway.CloudWatchLogsDestination`, inputting an arn is difficult.  We should fully expand the example to this:

```ts
    const loggroup = new cwlogs.LogGroup(this, "ApiGatewayAccessLogs", {});
    const restApi = new apigw.LambdaRestApi(this, 'books', {
      deployOptions: {
        accessLogDestination: new CloudWatchLogsDestination(loggroup),
        accessLogFormat: JSON.stringify({
          requestId: "$context.requestId",
          ip: "$context.identity.sourceIp",
          caller: "$context.identity.caller",
          user: "$context.identity.user",
          requestTime: "$context.requestTime",
          httpMethod: "$context.httpMethod",
          resourcePath: "$context.resourcePath",
          status: "$context.status",
          protocol: "$context.protocol",
          responseLength: "$context.responseLength"
        })
      }
    });
```

Let me know your thoughts :)

Author: You're right.
I forgot to revise the document.
The document has been modified, thank you!

  }
})
```
### Access Logging

> The following example will configure API Gateway to enable custom access logging.

Reviewer 2: Please address this comment.

#### CloudWatch Logs

```ts
const logGroup = new cwlogs.LogGroup(this, "ApiGatewayAccessLogs");
const api = new apigateway.RestApi(this, 'books', {
  deployOptions: {
    accessLogSetting: {

>         destinationArn: new apigateway.CloudWatchLogsDestination(logGroup),

Reviewer 2: Rename to `LogGroupDestination`.

        format: JSON.stringify({
            requestId: "$context.requestId",
            ip: "$context.identity.sourceIp",
            caller: "$context.identity.caller",
            user: "$context.identity.user",
            requestTime: "$context.requestTime",
            httpMethod: "$context.httpMethod",
            resourcePath: "$context.resourcePath",
            status: "$context.status",
            protocol: "$context.protocol",
            responseLength: "$context.responseLength"

>         })

Reviewer 2: We can do better here in terms of customer experience, and provide a more programming lang friendly and more a natural experience. Something like - 

```ts
format: {
  requestId: AuthorizerLogFormat.contextRequestId(),
  ip: AuthorizerLogFormat.contextIdentitySourceIp(),
  ...
}
```

or something along those lines.

The `JSON.stringify` can occur within the implementation of the construct, and hide it away from users.

Reviewer: I totally agree with the rest of the comments in this revision, but this one is a bit controversial. 

Specifically, the format can be [43](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#context-variable-reference) different context variables, those context variables can be arbitrary based on header names/parameter key names as well. The ordering of the keys is important in CLF/CSV format (also you can specify any format whatsoever, the above is just a clean json example). Due to it being an arbitrary string, that's specific to the customer's use case, I am unsure abstracting it away is a good idea, as it's dictating specific formats or ways.

Reviewer: as an example, this is a valid accesslog format I made up on the fly that should be accepted:
`requestId[$context.requestId]` which would output: `requestId[21123-4145fds-342fds-5fdss]` in your access logs.

Reviewer 2: > as an example, this is a valid accesslog format I made up on the fly that should be accepted:
> requestId[$context.requestId] which would output: requestId[21123-4145fds-342fds-5fdss] in your access logs.

If you used the code (I made up above), this would look like -

```ts
format: {
  requestId: `requestId[${AuthorizerLogFormat.contextRequestId()}]`
}
```

> Specifically, the format can be 43 different context variables, those context variables can be arbitrary based on header names/parameter key names as well.

I can see how bootstrapping 43 context variables and a bunch more stage variables, input variables, util variables can be hard. This is where we will have to get creative. We can start here -

```ts
AccessLogFormat.context('authorizer.principalId')
AccessLogFormat.input('params')
```

The ones that are most commonly used can then be added onto this and built up as necessary and over time - 

```
AccessLogFormat.CONTEXT_REQUEST_ID
// or
AccessLogFormat.contextRequestId()
// The latter probably since some of these can have additional suffixes like 'claims' and 'authorizer'
```

})
```

> The following example will configure API Gateway to enable custom access logging

Reviewer 2: Let's add a separate section for access logging - `### Access Logging`.

The access logging feature lets you generate access logs in different formats such as CLF (Common Log Format), JSON, XML, and CSV.
More info can be found [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-logging.html).

Access logs can be configured to be sent to a specific CloudWatch log group. This can either be configured either for all stages of the RestApi or to specific stages.
The following example will configure API Gateway to enable custom access logging.

```ts
const format = JSON.stringify({
  requestId: apigateway.AccessLogFormat.contextRequestId(),
  ip: apigateway.AccessLogFormat.contextIdentitySourceIp(),
  caller: apigateway.AccessLogFormat.contextIdentityCaller(),
  user: apigateway.AccessLogFormat.contextIdentityUser(),
  requestTime: apigateway.AccessLogFormat.contextRequestTime(),
  httpMethod: apigateway.AccessLogFormat.contextHttpMethod(),
  resourcePath: apigateway.AccessLogFormat.contextResourcePath(),
  status: apigateway.AccessLogFormat.contextStatus(),
  protocol: apigateway.AccessLogFormat.contextProtocol(),
  responseLength: apigateway.AccessLogFormat.contextResponseLength()

> });

Reviewer 2: Thanks for making these changes. I think we need to give this another round of treatment treatment and encapsulate the `JSON.stringify()` here. 

Looking at the examples and the blog posts around this feature, it seems that the most common use case here is to use JSON structure or CLF as the log format. The CDK should make this use case feel more natural. 

For JSON - 

```ts
{
  format: AccessLogFormat.json({
    requestId: AccessLogVariable.contextRequestId(),
    'request_time': AccessLogVariable.contextRequestTime(),
    ...
  });
```

For CLF  -

```ts
{
  format: AccessLogFormat.clf(),
}
```

This, of course, should allow for the possibility that the customer may want a completely different structure to their access log and so should allow - 

```ts
{
  format: AccessLogFormat.custom(`......`);
}
```

Reviewer 3: Once we have factory functions a la `AccessLogFormat.whatever(...)` we can add as many conveniences as we want. For example, I think I'm going to poke my eyes out if I have to write:

```ts
AccessLogFormat.json({
 requestId: apigateway.AccessLogFormat.contextRequestId(),
  ip: apigateway.AccessLogFormat.contextIdentitySourceIp(),
  caller: apigateway.AccessLogFormat.contextIdentityCaller(),
  user: apigateway.AccessLogFormat.contextIdentityUser(),
  requestTime: apigateway.AccessLogFormat.contextRequestTime(),
  httpMethod: apigateway.AccessLogFormat.contextHttpMethod(),
  // ...
});
```

Not to say we shouldn't have the `.json()` method. We *should*, but we *also* should have things like the following:

```ts
AccessLogFormat.jsonWithStandardFields({
  // Long live autocomplete!
  requestId: true,
  ip: true,
  requestTime: false,
});
```

```ts
// We prefix '$context.' to everything to keep it manageable
AccessLogFormat.jsonWithStandardFields('requestId', 'identity.sourceIp', 'requestTime', 'httpMethod');
```

Or whatever we come up with...

Reviewer 3: By the way I think it makes more sense to have the `.json()` factory function and the predefined field names `.contextRequestId()` on different objects.

How about:

```ts
AccessLogFormat.json(...)

AccessLogField.REQUEST_ID
```
?

Reviewer 2: > By the way I think it makes more sense to have the .json() factory function and the predefined field names .contextRequestId() on different objects.

Yip. My suggestion above was `AccessLogFormat` and `AccessLogVariable` but I'm fine with `AccesLogField` as well.

  }
})
```
### Access Logging

> You can set the format of the access log.
> More info can be found [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-logging.html).

Reviewer 2: ```suggestion
Access logging creates logs everytime an API method is accessed. Access logs can have information on
who has accessed the API, how the caller accessed the API and what responses were generated.
Access logs are configured on a Stage of the RestApi.

Access logs can be expressed in a format of your choosing, and can contain any access details, with a
minimum that it must include the 'requestId'. The list of  variables that can be expressed in the access
log can be found
[here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#context-variable-reference).

Read more at [Setting Up CloudWatch API Logging in API
Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-logging.html)
```

  }
})
```
### Access Logging

This is in addition to the detailed execution logs already provided by Amazon CloudWatch for API requests made to your APIs.
The access logging feature lets you generate access logs in different formats such as CLF (Common Log Format), JSON, XML, and CSV.
More info can be found [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-logging.html).

Access logs can be configured to be sent to a specific CloudWatch log group. This can be configured either for all stages of the RestApi or for a specific stage.
The following example will configure API Gateway to enable custom access logging.

```ts
// production stage
const prdLogGroup = new cwlogs.LogGroup(this, "PrdLogs");
const api = new apigateway.RestApi(this, 'books', {
  deployOptions: {

>     accessLogDestination: new apigateway.CloudWatchLogsDestination(prdLogGroup),

Reviewer 2: Update per the new name

Author: I'm sorry, I don't know what you want me to do.
I would like a few more comments.

Reviewer 2: Replace `CloudWatchLogsDestination` with `LogGroupLogDestination`

Author: I understand!

    accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields()
  }
})
const deployment = new apigateway.Deployment(stack, 'Deployment', {api});

// development stage
const devLogGroup = new cwlogs.LogGroup(this, "DevLogs");
new apigateway.Stage(this, 'dev', {
  deployment,
  accessLogDestination: new apigateway.LogGroupLogDestination(devLogGroup),
  accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields({
    caller: false,
    httpMethod: true,
    ip: true,
    protocol: true,
    requestTime: true,
    resourcePath: true,
    responseLength: true,
    status: true,
    user: true
  })
});
```

> The access log can be structured freely. The following will configure [CLF format](https://en.wikipedia.org/wiki/Common_Log_Format).

Reviewer 2: ```suggestion
The following code will generate the access log in the [CLF format](https://en.wikipedia.org/wiki/Common_Log_Format).
```

```ts
const logGroup = new cwlogs.LogGroup(this, "ApiGatewayAccessLogs");
const api = new apigateway.RestApi(this, 'books', {
  deployOptions: {
    accessLogDestination: new apigateway.LogGroupLogDestination(logGroup),
    accessLogFormat: apigateway.AccessLogFormat.clf();
  })
};

> ```

Reviewer 2: ````suggestion
```

You can also configure your own access log format by using the `AccessLogFormat.custom()` API.
`AccessLogField` provides commonly used fields. The following code configures access log to contain
 -

```ts
const logGroup = new cwlogs.LogGroup(this, "ApiGatewayAccessLogs");
new apigateway.RestApi(this, 'books', {
  deployOptions: {
    accessLogDestination: new apigateway.LogGroupLogDestination(logGroup),
    accessLogFormat: apigateway.AccessLogFormat.custom(
      `${AccessLogField.contextErrorMessage()} ${AccessLogField.contextErrorMessageString()}`);
  })
};
```

````

Reviewer 2: Please address this comment.
